[Tan Weiu Cheng] iP

DavidTan0527 opened the pull request:
# Nikki

> Rule 35 of JS: if it exists, there is a JavaScript framework with that name

Nikki is a *terminal-based application* for you to manage your tasks ~like a pro manager (jkjk)~

## Installation Guide
1. Download it from [here](https://github.com/DavidTan0527/ip/releases/tag/A-Jar).
2. Open your favorite terminal application.
3. Run `java -jar path/to/file.jar`
4. Enjoy! 🤗 

## Features
- Create, Read, Delete tasks
- Mark tasks as done or not
- Read and write to file for saving tasks in permanent storage
- 3 types of tasks: Deadline, Event, Todo
- Syntax checking and parsing for commands

## ToDos 🗒️ 
- [ ] Add command subtypes to `Command` for better tokenization
- [ ] Write command list for user guide

---

## Questions
*Note: this section is for nerds*

### Command tokenization
Currently, I'm using `regex` to parse my commands. For example, below is the regex for parsing the command `deadline`

```regex
deadline (.*) /(by) ((?:[1-9]|[12][0-9]|3[01])/(?:[1-9]|1[0-2])/[0-9]{4})
```

which is used as follows[^parser]

```java
Pattern p = Pattern.compile(/* regex above */);
Matcher m = p.matcher(cmd);

if (m.matches()) {
    if (m.groupCount() > 0) {
        String arg = m.group(1);
        HashMap<String, String> kwargs = new HashMap<>();

        // kwargs start at 2,
        // The last one is groupCount() - 1, its value being groupCount()
        for (int i = 2; i < m.groupCount(); i += 2) {
            kwargs.put(m.group(i), m.group(i + 1));
        }

        return new Command(cmdName, arg, kwargs);
    }
    else {
        return new Command(cmdName);
    }
} else {
    // Syntax error
    ...
}
```

The motivation of such design is for scaling for commands such that there can be more keyword arguments and it can be in arbitrary order.

**Question**: Is there a better way to achieve that result?

[^parser]: found in [`src/main/java/nikki/command/CommandParser.java`](https://github.com/DavidTan0527/ip/blob/master/src/main/java/nikki/command/CommandParser.java)

### File I/O handling
Since I'm reading from and writing to file, I would need to check its existence. My current solution is to create a file if it doesn't exist upon reading.

The caveat is when the user has ***no** write permission* in the directory of execution, then the file is not created, which causes issues when writing to the file. To mitigate this, I have also put the same file check + creation routine in the method that writes to the file.

However, it feels redundant to repeat the same procedure as it would've succeeded the first time and wouldn't need a second time. Alternatively, the second would've failed for the same reason as the first time.

**Question**: I'm not sure if my current approach[^file] is fine as it is or if there're any better solutions.

[^file]: found in [`src/main/java/nikki/Storage.java`](https://github.com/DavidTan0527/ip/blob/master/src/main/java/nikki/Storage.java)

Commit:
Add Gradle support

DavidTan0527 committed:
Greet, Echo, Exit

DavidTan0527 committed:
Implement add, list of tasks

DavidTan0527 committed:
Create parser to handle user input

DavidTan0527 committed:
Update parser to only parse known commands

DavidTan0527 committed:
Implement task mark/unmark

DavidTan0527 committed:
Add Todo, Event, Deadline, as types of Task

DavidTan0527 committed:
Update string representation for task types

DavidTan0527 committed:
Update command and command parser for better tokenization of args and kwargs

DavidTan0527 committed:
Add handler for todo, event, deadline task types

DavidTan0527 committed:
Minor change log format

DavidTan0527 committed:
Add automated text UI testing

DavidTan0527 committed:
Add DukeException class

DavidTan0527 committed:
Add exception handling

DavidTan0527 committed:
Implement delete task

DavidTan0527 committed:
Fix code styles and refactor code

DavidTan0527 committed:
Update Tasks string representation for file save

DavidTan0527 committed:
Fix tag field issue for Tasks

DavidTan0527 committed:
Add file support for saving and loading Tasks

DavidTan0527 committed:
Add datetime parsing to Deadline and Event

DavidTan0527 committed:
Merge 'branch-Level-7' with file save features

DavidTan0527 committed:
Merge conflicts in Deadline.java and Event.java

DavidTan0527 committed:
Fix Deadline and Event date save format

DavidTan0527 committed:
Rename variables

DavidTan0527 committed:
Refactor logging methods into ui class

DavidTan0527 committed:
Organize classes into packages

DavidTan0527 committed:
Add tests for TaskList and CommandParser

DavidTan0527 committed:
Add file existence check to storage save method

DavidTan0527 committed:
Update .gitignore to exclude jar related directory

DavidTan0527 committed:
Update javadoc in all classes

DavidTan0527 committed:
Change parameter name in Deadline constructor

DavidTan0527 committed:
Merge conflicts from branch-A-JavaDoc to master

DavidTan0527 committed:
Change method name to better describe behavior

DavidTan0527 committed:
Fix usage of renamed method

DavidTan0527 committed:
Merge conflicts from branch-A-CodingStandard to master

DavidTan0527 committed:
Fix usage of renamed method in test

DavidTan0527 committed:
Merge branch 'branch-A-CodingStandard'

DavidTan0527 committed:
Add command to find Task with keyword

DavidTan0527 committed:
Merge conflicts from branch-Level-9 to master

DavidTan0527 committed:
Update javadocs to conform with coding standard

DavidTan0527 committed:
Rename Duke to Nikki to be consistent with bot name choice

DavidTan0527 committed:
Merge branch 'master' into branch-A-Gradle

DavidTan0527 committed:
Configure gradle for checkstyle

DavidTan0527 committed:
Fix coding style according to gradle check

DavidTan0527 committed:
Merge branch 'branch-A-Gradle'

DavidTan0527 committed:
Update README

goel-a commented on src/main/java/nikki/task/Task.java:
```diff
@@ -0,0 +1,204 @@
+
+public class Task {
+    /** Mark in checkbox of completed Task */
+    private static final char completedMark = 'X';
```
Could use uppercase variable names to store constant

DavidTan0527 replied:
Thanks for pointing this out!

goel-a commented on src/main/java/nikki/task/Task.java:
```diff
@@ -0,0 +1,204 @@
+    private static final char completedMark = 'X';
+
+    /** Mark in checkbox of uncompleted Task */
+    private static final char incompleteMark = ' ';
```
Could use uppercase variable names to store constant

goel-a commented on src/main/java/nikki/task/Task.java:
```diff
@@ -0,0 +1,204 @@
+     *
+     * @return formatted string for saving Task
+     */
+    public String fileSaveFormat() {
```
Could use a verb for the name of the method

DavidTan0527 replied:
Will do, thanks!

goel-a commented on src/main/java/nikki/NikkiException.java:
```diff
@@ -0,0 +1,7 @@
+package nikki;
+
+public class NikkiException extends Exception {
+    public NikkiException(String errorMessage) {
```
Could consider adding a header comment to describe what kind of exceptions are being handled by this class/method

goel-a commented on src/main/java/nikki/task/TaskList.java:
```diff
@@ -0,0 +1,185 @@
+     *
+     * @return formatted string of Tasks for saving to file
+     */
+    public String tasksFileSaveFormat() {
```
Could consider naming the method using a verb

goel-a reviewed (commented):
Looks good! 
Some cosmetic changes recommended

DavidTan0527 committed:
Fix coding standards

DavidTan0527 committed:
Rename method in Task

DavidTan0527 committed:
Add JavaFX GUI

DavidTan0527 committed:
Merge branch 'branch-Level-10'

DavidTan0527 committed:
Update build.gradle for JavaFX bundling

DavidTan0527 committed:
Add assertions for non-empty response string builders

DavidTan0527 committed:
Refactor code to use guard statements and SLAP

DavidTan0527 committed:
Extract constant and use guard statements

DavidTan0527 committed:
Rename method

DavidTan0527 committed:
Merge pull request #1 from DavidTan0527/branch-A-CodeQuality

Adapt code to Code Quality standard

DavidTan0527 committed:
Add github actions workflow

DavidTan0527 committed:
Fix code quality and checkstyle issues

DavidTan0527 committed:
Merge branch 'branch-A-Assertions'

DavidTan0527 committed:
Implement fuzzy search for Tasks

DavidTan0527 committed:
Add acknowledgement in README

DavidTan0527 committed:
Fix similarity cutoff for task find

The command "find" used to search for task uses a fuzzy string matching
algorithm using Levenshtein distances between strings to calculate
similarities.

The chosen strategy of matching is a partial match, as to provide a
substring search as well. The output of the algorithm is between 0 and
100, hence a fix to the cutoff was required.

75 is chosen to be the threshold as it was experimented to be the most
forgiving while at the same time requires a "close enough" match.

DavidTan0527 committed:
Merge branch 'branch-C-BetterSearch'

DavidTan0527 committed:
Fix issue of gui not closing after bye command

DavidTan0527 committed:
Change to wider window

DavidTan0527 committed:
Change dialog box layout and color, and recolor main window

DavidTan0527 committed:
Merge branch 'branch-A-BetterGui'

DavidTan0527 committed:
Remove unused imports

DavidTan0527 committed:
Set theme jekyll-theme-hacker

DavidTan0527 committed:
Add UI screenshot for github pages

DavidTan0527 committed:
Add window title

DavidTan0527 committed:
Merge branch 'master' of github.com:DavidTan0527/ip

DavidTan0527 committed:
Change screenshot to include window title

DavidTan0527 committed:
Update user guide first draft

DavidTan0527 committed:
Fix broken link and adjust image style for user guide

DavidTan0527 committed:
Set theme jekyll-theme-midnight

DavidTan0527 committed:
Set theme jekyll-theme-slate

DavidTan0527 committed:
Quick fix markdown issue within accordions

DavidTan0527 committed:
Merge branch 'master' of github.com:DavidTan0527/ip

DavidTan0527 committed:
Wrap detail block in paragraph

DavidTan0527 committed:
Change image tag to markdown image

DavidTan0527 committed:
Remove jekyll include block and update feature section

DavidTan0527 committed:
Update feature and faq section on user guide

DavidTan0527 committed:
Change find command to case insensitive, and add clear command

DavidTan0527 committed:
Add response to clear command

DavidTan0527 committed:
Update user guide to reflect latest version
